Dockerfile

```diff
@@ -40,15 +41,293 @@ FROM alpine/terragrunt:latest as terragrunt
+## @generated by .automation/build.py using descriptor files, please do not update manually ##
+#############################################################################################
+
+ENV NODE_OPTIONS="--max-old-space-size=8192" \
+    NODE_ENV=production
+#NPM__START
+WORKDIR /node-deps
+RUN npm --no-cache install --ignore-scripts --omit=dev \
+                sfdx-cli \
+                typescript \
+                @coffeelint/cli \
+                jscpd \
+                stylelint \
+                stylelint-config-standard \
+                stylelint-config-sass-guidelines \
+                stylelint-scss \
+                gherkin-lint \
+                graphql \
+                graphql-schema-linter \
+                npm-groovy-lint \
+                htmlhint \
+                eslint \
+                eslint-config-airbnb \
+                eslint-config-prettier \
+                eslint-config-standard \
+                eslint-plugin-import \
+                eslint-plugin-jest \
+                eslint-plugin-node \
+                eslint-plugin-prettier \
+                eslint-plugin-promise \
+                eslint-plugin-vue \
+                @babel/core \
+                @babel/eslint-parser \
+                @microsoft/eslint-formatter-sarif \
+                standard \
+                prettier \
+                @prantlf/jsonlint \
+                eslint-plugin-jsonc \
+                v8r \
+                npm-package-json-lint \
+                npm-package-json-lint-config-default \
+                eslint-plugin-react \
+                eslint-plugin-jsx-a11y \
+                markdownlint-cli \
+                markdown-link-check \
+                markdown-table-formatter \
+                @stoplight/spectral-cli \
+                secretlint \
+                @secretlint/secretlint-rule-preset-recommend \
+                @secretlint/secretlint-formatter-sarif \
+                cspell \
+                sql-lint \
+                tekton-lint \
+                prettyjson \
+                @typescript-eslint/eslint-plugin \
+                @typescript-eslint/parser  && \
+    echo "Cleaning npm cache..." \
+    && npm cache clean --force || true \
+    && echo "Changing owner of node_modules files..." \
+    && chown -R "$(id -u)":"$(id -g)" node_modules # fix for https://github.com/npm/cli/issues/5900 \
+    && echo "Removing extra node_module files..." \
+    && rm -rf /root/.npm/_cacache \
+    && find . -name "*.d.ts" -delete \
+    && find . -name "*.map" -delete \
+    && find . -name "*.npmignore" -delete \
+    && find . -name "*.travis.yml" -delete \
+    && find . -name "CHANGELOG.md" -delete \
+    && find . -name "README.md" -delete \
+    && find . -name ".package-lock.json" -delete \
+    && find . -name "package-lock.json" -delete \
+    && find . -name "README.md" -delete
+WORKDIR /
+
+#NPM__END
+
+FROM scratch AS copy-collector
```

Comment:
I like that pattern!

Reply:
In case you were wondering going multiarch doesn't add much to the download payload. Instead of offering a bundle of all arches the registry offers a json file with a redirect to a different image per arch. That way the only additional download is a small json file and an additional round trip (Before: Fetch -> metadata -> download layers. After: Fetch -> pointer -> metadata -> download layers)

Reply:
For the download size I wasn't concerned. It's great work 😀

Reply:
> By reducing the layer count with tricks like this and doing the builds of things on images other than the final one (so you can remove the build-deps from the final image) I've managed to knock almost half the image size off the internal image I'm basing some of this off of.

That's really impressive! What build deps are you thinking about? I think rust must have them kept (the compiler is needed for one of the tools I think, but could keep only compiled binaries of cargo-built tools). There is only one go-based linter that needs the go apk package, everything else is already not present in main. For dotnet, dotnet format and other tools must have the full sdk installed 😑. What I would imagine could be a potential to get rid eventually in another PR (since we can iterate), is like GCC for Python package dependencies, or everything needed only to build Python packages, but not in use. (Is the separate stages solve this? Is the solution to build wheels and install from wheels of another stage?)